File: src/main/java/com/extendedae_plus/integration/RecipeViewer/jei/ModJeiPlugin.java
package com.extendedae_plus.integration.RecipeViewer.jei;

import com.extendedae_plus.ExtendedAEPlus;
import com.extendedae_plus.ae.definitions.upgrades.EntitySpeedCardItem;
import mezz.jei.api.IModPlugin;
import mezz.jei.api.JeiPlugin;
import mezz.jei.api.constants.VanillaTypes;
import mezz.jei.api.ingredients.subtypes.ISubtypeInterpreter;
import mezz.jei.api.ingredients.subtypes.UidContext;
import mezz.jei.api.registration.ISubtypeRegistration;
import mezz.jei.api.runtime.IJeiRuntime;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.item.ItemStack;
import org.jetbrains.annotations.NotNull;

@JeiPlugin
public class ModJeiPlugin implements IModPlugin {
    private static final ResourceLocation UID =
            ExtendedAEPlus.getLocation("jei_plugin");

    @Override
    public ResourceLocation getPluginUid() {
        return UID;
    }

    @Override
    public void onRuntimeAvailable(IJeiRuntime jeiRuntime) {
        JeiRuntimeProxy.setRuntime(jeiRuntime);
    }


    @Override
    public void registerItemSubtypes(ISubtypeRegistration registration) {
        // Register NBT-based subtype interpreter so JEI treats different multipliers as distinct items
        registration.registerSubtypeInterpreter(
                VanillaTypes.ITEM_STACK,
                com.extendedae_plus.init.ModItems.ENTITY_SPEED_CARD.get(),
                new ISubtypeInterpreter<>() {
                    @Override
                    public @NotNull Object getSubtypeData(@NotNull ItemStack ingredient, @NotNull UidContext context) {
                        // 返回你想让 JEI 区分子类型的数据，这里用 multiplier
                        return EntitySpeedCardItem.readMultiplier(ingredient);
                    }

                    @Override
                    public @NotNull String getLegacyStringSubtypeInfo(@NotNull ItemStack ingredient, @NotNull UidContext context) {
                        // 返回同样的值给旧接口兼容
                        return String.valueOf(EntitySpeedCardItem.readMultiplier(ingredient));
                    }
                }
        );

    }
}

File: src/main/kotlin/com/extendedae_plus/integration/impl/recipeViewer/jei/ModJeiPlugin.kt
package com.extendedae_plus.integration.impl.recipeViewer.jei

import com.extendedae_plus.ExtendedAEPlus
import com.extendedae_plus.common.init.EAEPItems
import com.extendedae_plus.common.registry.dataComponent.DataTickingCard
import com.fish.fishlib.util.extension.invoke
import mezz.jei.api.IModPlugin
import mezz.jei.api.JeiPlugin
import mezz.jei.api.constants.VanillaTypes
import mezz.jei.api.ingredients.subtypes.ISubtypeInterpreter
import mezz.jei.api.ingredients.subtypes.UidContext
import mezz.jei.api.registration.ISubtypeRegistration
import mezz.jei.api.runtime.IJeiRuntime
import net.minecraft.world.item.ItemStack

@JeiPlugin
class ModJeiPlugin : IModPlugin {
    override fun getPluginUid() = UID

    override fun onRuntimeAvailable(runtime: IJeiRuntime) {
        ViewerJei.runtime = runtime
    }

    override fun registerItemSubtypes(registration: ISubtypeRegistration) {
        registration.registerSubtypeInterpreter(
            VanillaTypes.ITEM_STACK,
            EAEPItems.CardTicking(),
            object : ISubtypeInterpreter<ItemStack> {
                override fun getSubtypeData(ingredient: ItemStack, context: UidContext) =
                    DataTickingCard.fromStack(ingredient)

                @Deprecated("Deprecated in Java")
                override fun getLegacyStringSubtypeInfo(ingredient: ItemStack, context: UidContext) =
                    DataTickingCard.fromStack(ingredient).toString()
            }
        )
    }

    companion object {
        private val UID = ExtendedAEPlus.getLocation("jei_plugin")
    }
}
